4/lib/model.d.ts

```diff
@@ -0,0 +1,2422 @@
+   * must be true. Not affected by underscored setting.
+   */
+  updatedAt?: string | boolean;
+
+  /**
+   * Defaults to pluralized model name, unless freezeTableName is true, in which case it uses model name
+   * verbatim
+   */
+  tableName?: string;
+
+  /**
+   * Provide getter functions that work like those defined per column. If you provide a getter method with
+   * the
+   * same name as a column, it will be used to access the value of that column. If you provide a name that
+   * does not match a column, this function will act as a virtual getter, that can fetch multiple other
+   * values
+   */
+  getterMethods?: ModelGetterOptions;
+
+  /**
+   * Provide setter functions that work like those defined per column. If you provide a setter method with
+   * the
+   * same name as a column, it will be used to update the value of that column. If you provide a name that
+   * does not match a column, this function will act as a virtual setter, that can act on and set other
+   * values, but will not be persisted
+   */
+  setterMethods?: ModelSetterOptions;
+
+  /**
+   * Provide functions that are added to each instance (DAO). If you override methods provided by sequelize,
+   * you can access the original method using `this.constructor.super_.prototype`, e.g.
+   * `this.constructor.super_.prototype.toJSON.apply(this, arguments)`
+   */
+  instanceMethods?: Object;
+
+  /**
+   * Provide functions that are added to the model (Model). If you override methods provided by sequelize,
+   * you can access the original method using `this.constructor.prototype`, e.g.
+   * `this.constructor.prototype.find.apply(this, arguments)`
+   */
+  classMethods?: Object;
+
+  schema?: string;
+
+  /**
+   * You can also change the database engine, e.g. to MyISAM. InnoDB is the default.
+   */
+  engine?: string;
+
+  charset?: string;
+
+  /**
+   * Finaly you can specify a comment for the table in MySQL and PG
+   */
+  comment?: string;
+
+  collate?: string;
+
+  /**
+   * Set the initial AUTO_INCREMENT value for the table in MySQL.
+   */
+  initialAutoIncrement?: string;
+
+  /**
+   * An object of hook function that are called before and after certain lifecycle events.
+   * The possible hooks are: beforeValidate, afterValidate, beforeBulkCreate, beforeBulkDestroy,
+   * beforeBulkUpdate, beforeCreate, beforeDestroy, beforeUpdate, afterCreate, afterDestroy, afterUpdate,
+   * afterBulkCreate, afterBulkDestory and afterBulkUpdate. See Hooks for more information about hook
+   * functions and their signatures. Each property can either be a function, or an array of functions.
+   */
+  hooks?: HooksOptions;
+
+  /**
+   * An object of model wide validations. Validations have access to all model values via `this`. If the
+   * validator function takes an argument, it is asumed to be async, and is called with a callback that
+   * accepts an optional error.
+   */
+  validate?: ModelValidateOptions;
+
+}
+
+export abstract class Model {
```

Comment:
That is not possible either unfortunately, because static members cannot reference type arguments.
To be honest, I found it super ugly to have to write a interface that references itself in the generic type argument like you have to with Seq 3 typings:
```ts
interface Thing { // interface used to create/update an instance
  id?: number;
  name?: string;
}
// ThingInstance references itself in it's declaration 😨
interface ThingInstance extends Sequelize.Instance<ThingInstance, Thing> {
  id: number; // have to duplicate all the properties here so they are not optional 😩
  // you should add all instance methods here
  doSomething(): any;
}
```
And there wasn't even a way to type static `classMethods` without casting...
Passing `TAttributes` to be used for `update()` and `create()` wasn't 100% correct either. For `update()` all attributes are optional, while for `create()`, only those that have `allowNull: true` or a `defaultValue` are optional. So if you're gonna allow to pass a generic type argument for using those, you should actually allow to pass two, so you get compile-time errors for `create()`.
I really feel like an abstract class is the best type for `Model` because it actually _is_ an ES6 class in sequelize codebase now. The `abstract` keyword nicely protects against instantiating it without subclassing, which is not possible with interfaces. The generic thing made sense before because `Model` really was a generic class that returned a specific type of instance. But now it is just a class.

Using interfaces would not help. First, you would have to seperate static and instance interfaces, because interfaces cannot contain static members. Say we have this definition:
```ts
interface ModelInstance {
  save(): Promise<void>;
}
interface Model<TInstance> {
  new (): TInstance;
  findOne(): Promise<TInstance>;
}
```
If you try to extend it with `class User extends Model {}` you would get an error _Cannot find name 'Model'_, because `Model` is only an interface. So you would have to declare a variable with the type of the Model class:
```
const Model: Model;
```
But then you get _Generic type 'TestModel<TInstance>' requires 1 type argument(s)._, which means you have to type `const Model: Model<any>`, at which point you lost the ability to pass the type argument to `Model` when declaring your subclass.

There's still some time until Sequelize 4 is released and I hope TypeScript will adress [#5863](https://github.com/Microsoft/TypeScript/issues/5863) soon. Until then, the casting is not _that_ ugly really, see https://github.com/felixfbecker/typed-sequelize/blob/v4/4/test/usage.ts.
I would rather work with a real abstract class (that shows up as a class in my IDE and not two interfaces + one constant) and write `as User` behind my queries than have to pass x type arguments everywhere. It's not perfect, but polymorphic `this` would really be the correct syntax to type this, until we don't have this let's go with casting. The only thing I would make generic is `define()` since it is a factory function.